### PR TITLE
Add basic failover functionality

### LINKED ISSUE #733: RPC Redundancy/Failover Configuration
***Is your feature request related to a problem? Please describe.***
 Current RPC providers can have downtime, temporary connectivity issues, or rate limits that make clients transactions fail. Over the past year we have observed several windows of RPC failure, which could have been mitigated if near-api-js had configurations for multiple RPC providers.

***Describe the solution you'd like***
 Similar to issues #703 and #717, the core JSON RPC provider should be refactored to not only have retries, but allow for retries against multiple Providers. The default provider list can configure both near foundation and openshards RPC services. Provider list can be a single node string for backward compatibility OR an array of node strings.
 Up for discussion but needed: Create a failover threshold of retries for each provider, and a threshold for provider failures before defaulting to a different priority.

***Configuration Example:***
 ```bash
 RPC_MAINNET_PROVIDERS="https://rpc.mainnet.near.org,[https://mainnet-rpc.openshards.io](https://mainnet-rpc.openshards.io)"

RPC_GUILDNET_PROVIDERS="https://guildnet-rpc.openshards.io,[https://rpc.guildnet.near.org](https://rpc.guildnet.near.org)"
 ```

NOTE: Because near-api-js is used in many dapps and repos, this functionality is very key toward providing the easiest way to allow clients to decentralize their RPC access. This is critical for community attacks against public resources.

***Describe alternatives you've considered***
 Users must create multiple instances of the Near module with different providers configured and detect TXN failures. Not idea at all.

***Additional context***
 There is an ongoing effort to create a decentralized RPC for mainnet & guildnet using many of the openshards.io nodes with a redundant load-balancer.

### PR DESCRIPTION
**Pull request by gtsonevv**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/near/near-api-js/blob/master/CONTRIBUTING.md).
- [x] Commit messages follow the [conventional commits](https://www.conventionalcommits.org/) spec
- [x] **If this is a code change**: I have written unit tests.
- [x] **If this changes code in a published package**: I have run `pnpm changeset` to create a `changeset` JSON document appropriate for this change.
- [x] **If this is a new API or substantial change**: the PR has an accompanying issue (closes [#733](https://github.com/near/near-api-js/issues/733)) and the maintainers have approved on my working plan.

## Motivation

Added support for multiple API Keys (since we can have multiple RPC Servers)
Added basic failover functionality
With this design of failover, we will change (rotate) our server when needed and stick to this new server for our next calls. It will help to avoid long responses when first (or more) servers are down.

For now, server rotation is happening only on Timeout error.

Also, check this comment: https://github.com/near/near-api-js/issues/733#issuecomment-998869162

## Test Plan

Added unit tests.

## Related issues/PRs

https://github.com/near/near-api-js/issues/733
https://github.com/near/near-api-js/pull/760

### CODE REVIEW

File: packages/cookbook/api-keys/near-connection.js [hunk 1]
```diff
@@ -7,16 +7,16 @@ const CREDENTIALS_DIR = ".near-credentials";
...
 const ACCOUNT_ID = '<Replace this string with existing account ID>';
 
 const config = {
     networkId: 'testnet',
     keyStore,
-    nodeUrl: RPC_API_ENDPOINT,
-    headers: { 'x-api-key': API_KEY },
+    nodeUrl: '<Replace this string with your RPC server URL>',
+    // RPC server URL in apiKeys should match the one specified in nodeUrl
+    apiKeys: { '<Replace this string with your RPC server URL>': API_KEY },
```
Comment (vikinatora) [inline packages/cookbook/api-keys/near-connection.js:19]: If no config is provided, then the config won't be working:
`'<Replace this string with your RPC server URL>': API_KEY` is not a valid api key

Comment (vikinatora) [reply packages/cookbook/api-keys/near-connection.js:19]: Or is 'x-api-key' just an example?

File: packages/cookbook/api-keys/provider-example.js [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 
-const RPC_API_ENDPOINT = '<Replace this string with your RPC server URL>';
 const API_KEY = '<Replace this string with your API KEY>';
 
 const provider = new providers.JsonRpcProvider({
-    url: RPC_API_ENDPOINT,
-    headers: { 'x-api-key': API_KEY },
+    url: '<Replace this string with your RPC server URL>',
+    // RPC server URL in apiKeys should match the one specified in url
+    apiKeys: { '<Replace this string with your RPC server URL>': API_KEY }
```
Comment (vikinatora) [inline packages/cookbook/api-keys/provider-example.js:9]: Same comment as above

File: packages/providers/src/fetch_json.ts [hunk 3]
```diff
@@ -24,36 +26,40 @@ export async function fetchJson(connectionInfoOrUrl: string | ConnectionInfo, js
     } else {
         connectionInfo = connectionInfoOrUrl as ConnectionInfo;
     }
-
+    const currentRpcServer = typeof connectionInfo.url === 'string' ? connectionInfo.url : connectionInfo.url[0];
```
Comment (vikinatora) [inline packages/providers/src/fetch_json.ts:34]: What's the chance that `connectionInfo.url[0]` will be null? 🤔

Comment (vikinatora) [reply packages/providers/src/fetch_json.ts:34]: Maybe we should add validations for the provided config in the constructor

File: packages/providers/src/json-rpc-provider.ts [hunk 4]
```diff
@@ -333,6 +333,19 @@ export class JsonRpcProvider extends Provider {
         return await this.sendJsonRpc('gas_price', [blockId]);
     }
 
+    /**
+     * Part of the RPC failover design.
+     * Changing current (first) rpc server and moves it to the and of the queue.
```
Comment (vikinatora) [inline packages/providers/src/json-rpc-provider.ts]:
```suggestion
     * Changing current (first) rpc server and moves it to the end of the queue.
```

File: packages/providers/src/json-rpc-provider.ts [hunk 5]
```diff
@@ -333,6 +333,19 @@ export class JsonRpcProvider extends Provider {
...
+     */
+    private rotateRpcServers() {
+        if (typeof this.connection.url === 'string') { return; }
+        const currentRpcServer = this.connection.url.shift();
+        this.connection.url.push(currentRpcServer);
+        if (!process.env['NEAR_NO_LOGS']) {
+            console.warn(`Switched from ${currentRpcServer} RPC Server to ${this.connection.url[0]}`);
+        }
```
Comment (vikinatora) [inline packages/providers/src/json-rpc-provider.ts:346]: If `NEAR_NO_LOGS` doesn't exist -> it won't log. Shouldn't it be the other way around?

File: packages/providers/test/providers.test.js [hunk 6]
```diff
@@ -187,3 +187,19 @@ test('near json rpc fetch node status', async () => {
...
+    expect(provider.connection.url.length).toEqual(3);
+    expect(provider.connection.url[0]).toMatch(SERVER_1);
+    expect(provider.connection.url[1]).toMatch(SERVER_2);
+    expect(provider.connection.url[2]).toMatch(SERVER_3);
+    provider.rotateRpcServers();
+    expect(provider.connection.url.length).toEqual(3);
+    expect(provider.connection.url[0]).toMatch(SERVER_2);
+    expect(provider.connection.url[1]).toMatch(SERVER_3);
+    expect(provider.connection.url[2]).toMatch(SERVER_1);
+});
```
Comment (vikinatora) [inline packages/providers/test/providers.test.js:232]: Add a bad case test which contains bad configuration. The error should be caught at construction time I believe.

File: packages/wallet-account/src/near.ts [hunk 7]
```diff
@@ -52,9 +52,15 @@ export interface NearConfig {
...
+     * You can set multiple RPC Server URLs to turn on failover functionality.
      * @see {@link "@near-js/providers".json-rpc-provider.JsonRpcProvider | JsonRpcProvider}
      */
-    nodeUrl: string;
+    nodeUrl: string | string[];
+
+    /**
+     * RPC API Keys. Used to authenticate users on RPC Server.
+     */
+    apiKeys: string;
```
Comment (vikinatora) [inline packages/wallet-account/src/near.ts]: Shouldn't this be `{ [url: string]: string };`

File: packages/providers/src/json-rpc-provider.ts (hunk 5 above)
Comment (gtsonevv) [reply packages/providers/src/json-rpc-provider.ts:346]: there should be no logs if `NEAR_NO_LOGS` is `true`

File: packages/cookbook/api-keys/near-connection.js (hunk 1 above)
Comment (gtsonevv) [reply packages/cookbook/api-keys/near-connection.js:19]: `'<Replace this string with your RPC server URL>'` must be a custom RPC server URL and `API_KEY` a corresponding key for the URL

File: packages/providers/test/providers.test.js (hunk 6 above)
Comment (gtsonevv) [reply packages/providers/test/providers.test.js:232]: will add

File: packages/providers/src/fetch_json.ts [hunk 8]
```diff
@@ -24,36 +26,45 @@ export async function fetchJson(connectionInfoOrUrl: string | ConnectionInfo, js
     } else {
         connectionInfo = connectionInfoOrUrl as ConnectionInfo;
     }
+    
+    if (!connectionInfo.url && !connectionInfo.url.length && !connectionInfo.url[0]) {
+        throw new Error('RPC Server URL or the prioritized array of such URLs should not be empty');
+    }
```
Comment (vikinatora) [inline packages/providers/src/fetch_json.ts]: Additionally, we should check if all of the provided urls are of type string i believe

File: packages/providers/src/json-rpc-provider.ts [hunk 9]
```diff
@@ -374,6 +387,8 @@ export class JsonRpcProvider extends Provider {
             } catch (error) {
                 if (error.type === 'TimeoutError') {
                     Logger.warn(`Retrying request to ${method} as it has timed out`, params);
+                    // switch  to another server if it's available
```
Comment (vikinatora) [inline packages/providers/src/json-rpc-provider.ts]: nitpick
```suggestion
                    // switch to another server if it's available
```

**Pull request closed without merging**